HeterogeneousCore/AlpakaInterface/interface/SimpleVector.h

```diff
@@ -0,0 +1,142 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_SimpleVector_h
+#define HeterogeneousCore_AlpakaInterface_interface_SimpleVector_h
+
+//  author: Felice Pantaleo, CERN, 2018
+//  alpaka integration in cmssw: Adriano Di Florio, 2022
+
+#include <type_traits>
+#include <utility>
+#include <alpaka/alpaka.hpp>
+
+namespace cms::alpakatools {
+
+  template <class T>
+  struct SimpleVector {
+    constexpr SimpleVector() = default;
+
+    // ownership of m_data stays within the caller
+    constexpr void construct(int capacity, T *data) {
+      m_size = 0;
+      m_capacity = capacity;
+      m_data = data;
+    }
+
+    inline constexpr int push_back_unsafe(const T &element) {
+      auto previousSize = m_size;
+      m_size++;
+      if (previousSize < m_capacity) {
+        m_data[previousSize] = element;
+        return previousSize;
+      } else {
+        --m_size;
+        return -1;
+      }
+    }
+
+    template <class... Ts>
+    constexpr int emplace_back_unsafe(Ts &&...args) {
+      auto previousSize = m_size;
+      m_size++;
+      if (previousSize < m_capacity) {
+        (new (&m_data[previousSize]) T(std::forward<Ts>(args)...));
+        return previousSize;
+      } else {
+        --m_size;
+        return -1;
+      }
+    }
+
+    ALPAKA_FN_ACC inline T &back() { return m_data[m_size - 1]; }
+
+    ALPAKA_FN_ACC inline const T &back() const {
+      if (m_size > 0) {
+        return m_data[m_size - 1];
+      } else
+        return T();  //undefined behaviour
+    }
+
+    // thread-safe version of the vector, when used in a CUDA kernel
+    template <typename TAcc>
+    ALPAKA_FN_ACC int push_back(const TAcc &acc, const T &element) {
+      auto previousSize = alpaka::atomicAdd(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
```

Comment:
do we have `alpaka::atomicInc(...)` instead of `alpaka::atomicAdd(..., 1)` ?

Reply:
I assume you meant

> will _not_ make the code any simpler.

While the code will not be simpler, at least on CUDA the performance of `atomicInc` is potentially better than that of `atomicAdd`, because the operations on the same variable in the same warp are implemented using warp primitives, avoiding the contention.

Reply:
Actually, I have to correct myself: looking at the output of compiler explorer for `atomicInc(ptr, 0xffffffff)` and `atomicAdd(ptr, 1)`, they generate exactly the same SASS code, both for global and shared memory.

So feel free to use the one that you prefer.